### PR TITLE
Configuration to automatically notify producers on order cycle close

### LINKED ISSUE #8529: Automatically email producers
## Description

As a Hub I want to have automatic notifications to my producers at the end of every order cycle.

I want to enable this in 'Edit Order Cycle' -> General Settings.
This should be a checkbox with the text 'Automatically notify producers with their orders via emails when order cycles close.'

## Acceptance Criteria & Tests

1. The email that is sent should be the same as the 'Notify Producers' email
2. The email should be sent to the Producers and CC the Hubs
3. The email should be sent at the time that the Order Cycle closes
4. The text in the Order Cycle page should be translated.

_______________________________________________________________________________________

## Original Post

Another request from UK hubs; an option to make the notifications emails at the end of an order cycle send automatically when the order cycle closes, so that producers will receive them by the next morning after close.

Could be something like a checkbox in the order cycle.

Also, there isn't anything that shows whether or not the emails have been sent for an order cycle, which may cause issues if there are multiple people doing hub admin, so a Sent/Not Sent flag would be useful.

### PR DESCRIPTION
**Pull request by seballot**
#### What? Why?

Closes #8529

@Matt-Yorkley did the background job part
I have just added the config in the form (in advanced settings)

I also adjusted the notify producer button to reflect wether or not the emails have been already sent

image

image

Is it clear enough for you?

### How to test?

- Create an OC, set advance setting "automatic notification" checkbox to true
- Wait for the OC to close and you should receive the producer emails
- Go back to OC and the `notify producer` button should be updated with green checkbox
- Change the close date to a future data (like if you were reusing same OC for another data)
- Update the page, the `notify producer` button should be back to normal
- Wait for the OC to close again and you should receive the producer emails

### CODE REVIEW

**Review by jibees — APPROVED** [review]
LGTM thanks.
I'd add some _How to test?_ as you know :p (sorry)

**Review by Matt-Yorkley — COMMENTED** [review]
Nice. I think this feature will need a bit more work though. 

Users can change Order Cycles and (for example) set new dates for a different week (from the order cycles page, or in the order cycle edit page). In that case we should probably clear the `processed_at` field (reset it to `nil`) otherwise the modified order cycle won't send emails the next time the cycle closes.

Maybe we could do something like a `before_save` callback on OrderCycle which checks if the `:orders_close_at` field has changed, and if so, sets `:processed_at` to `nil`?

File: app/assets/stylesheets/admin/advanced_settings.scss [hunk 1]
```diff
@@ -4,18 +4,15 @@
...
-    padding: 20px 0px;
-
-    .column.alpha, .columns.alpha {
-      padding-left: 20px;
-    }
-
-    .column.omega, .columns.omega {
-      padding-right: 20px;
-    }
+#toggle_settings {
```
[the remark this answers is not in the document]
Comment (seballot) [reply app/assets/stylesheets/admin/advanced_settings.scss:10]: Hi ! the button with the id `toggle_settings` was already there, it's the one for toggling advance details panel `app/views/admin/order_cycles/_advanced_settings.html.haml`
I just improved the display

**Comment by seballot** [comment]
Hi @Matt-Yorkley !

Nice idea, I have implemented it, adding the condition that the new closing date should be later that previous, otherwise it could notify twice an OC for nothing. Do you agree?

**Comment by Matt-Yorkley** [comment]
Thanks @seballot yeah that looks great :+1:

There's a failing spec now at `spec/models/spree/order_spec.rb:1258` which indicates the `orders_close_at > orders_close_at_was` comparison will throw a fatal error if `orders_close_at` is `nil`. Maybe it needs a guard clause in there for that case?

**Review by Matt-Yorkley — APPROVED** [review]

File: spec/models/order_cycle_spec.rb [hunk 2]
```diff
@@ -542,6 +545,30 @@
...
+
+    it "reset processed_at if close date change in future" do
+      expect(oc.processed_at).to_not be_nil
+      oc.update!(orders_close_at: 1.week.from_now)
+      expect(oc.processed_at).to be_nil
+    end
+
+    it "it does not reset processed_at if close date change in the past" do
+      expect(oc.processed_at).to_not be_nil
+      oc.update!(orders_close_at: 2.day.ago)
```
Comment (Matt-Yorkley) [inline spec/models/order_cycle_spec.rb:561]:
```suggestion
      oc.update!(orders_close_at: 2.days.ago)
```

Comment (Matt-Yorkley) [reply spec/models/order_cycle_spec.rb]: Fixes a quick Rubocop violation :ok_hand:

**Comment by filipefurtad0** [comment]
Hey @seballot ,
Between code-review and testing, some conflicts arose; could you please rebase this PR?
Thanks!

**Comment by seballot** [comment]
Hi @filipefurtad0, sure, it's done

**Comment by filipefurtad0** [comment]
Hey @seballot ,

this works very nicely :muscle: 

Before this PR the advanced settings looked like this:
image

After your PR:
image


Your steps to reproduce lead to the expected outcome:

- after setting the automatic notification on the Advanced Settings menu
- the green check appears on the "Notify Producers" button, once the OC closes; "Notify Producers" changes to "Re-Notify Producers".
- the emails arrive, automatically
- Clicking the "Re-Notify Producers" resends the email, as expected.
- Setting the OC close time to a future date removes the green check from the button.

image


This looks great, thanks!
Merging.

**Comment by Matt-Yorkley** [comment]
Ah we'll want @openfoodfoundation/train-drivers-product-owners to look at this, as it's a new feature that warrants user guide updates etc...

**Comment by lin-d-hop** [comment]
I'm super excited about this feature! I've literally employed someone to press buttons on a spreadsheet because this didn't exist :facepalm: 

And yes to the user guide update. Though obviously the UG update is not a blocker to release.

**Comment by RachL** [comment]
@filipefurtad0 was this PR tested with the "single producer" OC version? I can't seem to receive the notification in this case but I could be wrong. A second check would be welcomed.

Also once the OC is closed or even just saved the checkbox remains unchecked when you navigate the advance section a second time. Which is confusing when you want to check whether or not you have checked the box.

I would like to open 2 new issue to improve this: 

1. keep the checkbox checked if it was checked once.

2. If the checkbox is checked. Keep it checked when cloning the OC.

@lin-d-hop @filipefurtad0 what do you think?

**Comment by filipefurtad0** [comment]
Hey @RachL ,

It was tested for a OC in which both the coordinator and another producer acted as suppliers; both got the email during the tests, but if this is not happening on your setup then it's definitively worth a second look. I'll investigate and get back to you on this.

**Comment by seballot** [comment]
Hi @RachL 

All the things you mention were working okay at the time of implementing it (checkbox still checked, cloning OC) 

Maybe the rebase I did recently introduced some error? or maybe something is wrong with your setup?

**Comment by filipefurtad0** [comment]
Ok, it took me a bit, but it seems to work as well on a different setup / staging server:

image

This was verified on two different scenarios:
- actively closing the OC (setting a past date) 
- waiting till the OC closes

I think we're good, but indeed, I noticed that we do need to wait a couple of minutes. I'd think this is fine, what do you think @RachL? I guess in any case it migth be a relevant detail to include in the user-guide.

**Comment by filipefurtad0** [comment]
There is something though which might need attention - namely before this weeks deployment:

@jibees spotted this in a developer environment, on the current master branch - I can reproduce it as well in my dev env:

image

This error appears when clicking the pencil-symbol, to edit a given order-cycle. However, I can't see to reproduce this in staging. There are differences between dev / staging / production environments, but I'm not 100% sure this is the cause.

Any advice or hint on the scenario to reproduce it in our servers? Ping @seballot .

**Comment by Matt-Yorkley** [comment]
@filipefurtad0 did you run `bundle exec rake db:migrate` in your dev environment?

**Comment by filipefurtad0** [comment]
Perfect - that's it. Thanks @Matt-Yorkley !

**Pull request merged**